_posts/2021-01-12-feedback-to-users-requests-in-distributed-systems.md

```diff
@@ -0,0 +1,105 @@
+On the other hand, users are not scaled out into user instances. A user is a single user, in which case it's enough to store the correlation ID as a cookie or in the browser's local storage. It doesn't need to be replicated. In a more advanced scenario where the user can start a request on device A and finish it on device B, a physically different device, the correlation ID can be stored in the user's profile on the server. Still, it's a single user.
+
+This being said, the fact that web servers can be scaled out, and users are single instances opens up another interesting problem to handle. In most cases, we don't want to use polling techniques to check the status of a request; this is because of polling's inefficiency. Imagine a request that takes 10 minutes to be fulfilled and a polling interval of 10 seconds; we're going to "annoy" the backend infrastructure 59 times before getting the answer we're looking for. Technology advances, and nowadays, the de facto solution is to use WebSockets. It's the web server's job to call you back with information, instead of you continually pinging them for information. Our sample scenario means that when the backend system completes the request processing, it sends a notification to the user device, e.g., the browser, using a WebSocket message. Web servers are scaled out into instances, and users are not. WebSockets are linked to users, not to web servers. Let's see what the implications of this relationship are:
+
+- The web server accepted the user request and, after doing all the ceremonies, sends a message to the backend service
+- The backend service completes the request processing and sends back a response message to the webserver.
+
+> Q: To which webserver instance?
+
+If web servers are scaled out and use a queue to communicate with backend services, there is no guarantee that the backend service response message goes to the same web server instance that sent out the initial request. If there was no user, and no WebSockets, involved we had no problems, scaled out instances can be designed to be stateless, and thus, what instance handles a message is not that important. A device using WebSocket connects directly to a specific web server instance. In such a scenario, if the user sends the initial request to web server instance A, and web server instance B is handling the response message from the backend service, there is no user's WebSocket connected to instance B. The status is lost because web server instance B is missing some context or information to talk back to the user's device.
```

Comment:
```suggestion
If web servers are scaled out and use a queue to communicate with backend services, there is no guarantee that the backend service response message goes to the same web server instance that sent out the initial request. If there was no user, and no WebSockets, involved we had no problems; scaled out instances can be designed to be stateless, and thus, the particular instance that handles a message is not important. But a device using WebSockets connects directly to a specific web server instance. In this scenario, if the user sends the initial request to web server instance A, and web server instance B is handling the response message from the backend service, there is no user's WebSocket connected to instance B. The status is lost because web server instance B is missing some context or information to talk back to the user's device.
```